### PR TITLE
Pianoroll2

### PR DESCRIPTION
**Pull request by blackears**
Resolves: *(direct link to the issue)*

*(short description of the changes and the motivation to make the changes)*

- [ ] I signed [CLA](https://musescore.org/en/cla)
- [ ] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [ ] I made sure the code compiles on my machine
- [ ] I made sure there are no unnecessary changes in the code
- [ ] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [ ] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [ ] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

File: src/pianoroll/view/pianorollview.h [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    void cutChord(const QPointF& pos);
+    void eraseNote(const QPointF& pos);
+    Ms::Fraction roundDownToSubdivision(double wholeNote);
+
+    void pasteNotes(const QString& copiedNotes, Ms::Fraction pasteStartTick, Ms::Fraction lengthOffset, int pitchOffset, bool xIsOffset);
+    std::vector<Ms::Note*> addNote(Ms::Fraction startTick, Ms::Fraction duration, int pitch, int track);
+    bool cutChordRest(Ms::ChordRest* targetCr, int track, Ms::Fraction cutTick, Ms::ChordRest*& cr0, Ms::ChordRest*& cr1);
+    std::vector<Ms::Note*> getSegmentNotes(Ms::Segment* seg, int track);
+
+    static void append(std::vector<Ms::Note*>& a, std::vector<Ms::Note*>& b)
```
Comment (wizofaus) [inline src/pianoroll/view/pianorollview.h:187]: Questionable whether this function belongs here, but if you make the second parameter a const ref you should avoid most of those build errors.

File: src/pianoroll/view/pianorollview.h [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+public:
+    enum class PianorollTool : char { SELECT, ADD, CUT, ERASE };
+    Q_ENUM(PianorollTool)
+
+    static const BarPattern barPatterns[];
+
+private:
+    INJECT(pianoroll, context::IGlobalContext, globalContext)
+    INJECT(pianoroll, IPianorollController, controller)
+    INJECT(playback, audio::IPlayback, playback)
```
Comment (wizofaus) [inline src/pianoroll/view/pianorollview.h:83]: believe it should be pianoroll for the 1st parameter here (not entirely sure what difference it makes though!)

File: src/pianoroll/view/pianorollview.h (hunk 1 above)
Comment (blackears) [reply src/pianoroll/view/pianorollview.h]: I've rewritten the code to avoid this method entirely.

File: src/pianoroll/view/pianorollview.h (hunk 2 above)
Comment (blackears) [reply src/pianoroll/view/pianorollview.h]: Fixed.

Comment (igorkorsukov) [reply src/pianoroll/view/pianorollview.h]: > not entirely sure what difference it makes though!

It's actually just for information. In the future, I want to make a diagnostic page on which to show what depends on what.

**Comment by blackears** [comment]
Created new PR to replace this one. https://github.com/musescore/MuseScore/pull/9109/files

**Pull request closed without merging**